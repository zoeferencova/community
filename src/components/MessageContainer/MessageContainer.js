import React, { Component } from "react";
import ChatService from "../../services/chat-service";
import CommUnityContext from "../../contexts/context";
import MessageSideBar from "../MessageSideBar/MessageSideBar";
import MessageHeading from "../MessageHeading/MessageHeading";
import MessageInput from "../MessageInput/MessageInput";
import Messages from "../Messages/Messages";
import { MESSAGE_SENT, TYPING } from "../../message-utils/events";
import styles from "../MessageLayout/MessageLayout.module.css";

export default class MessageContainer extends Component {
    static contextType = CommUnityContext;

    constructor(props) {
        super(props);
        this._isMounted = false;
        this.state = {
            chats: [],
            mobileDisplay: null
        }
    }

    componentDidMount() {
        this._isMounted = true;
        this.setState({ chats: this.context.chats, mobileDisplay: this.context.activeChat === null ? "contacts" : "chats" })
    }

    componentWillUnmount() {
        this._isMounted = false;
    }

    updateTypingInChat = (chatId) =>{
		return ({isTyping, user})=>{
			if (user.id !== this.context.user.id) {

				const { chats } = this.state

				let newChats = chats.map((chat) => {
					if (chat.id === chatId) {
                        const userIds = chat.typingUsers.map(typingUser => typingUser.id)
						if (isTyping && !userIds.includes(user.id)) {
							chat.typingUsers.push(user)
						} else if(!isTyping && userIds.includes(user.id)) {
							chat.typingUsers = chat.typingUsers.filter(u => u.id !== user.id)
						}
					}
					return chat
				})
				this._isMounted && this.setState({ chats: newChats })
			}
		}
	}

    setActiveChat = activeChat => {
        this.context.updateActiveChat(activeChat.id)
    }

    mobileDisplayContacts = () => {
        this.setState({ mobileDisplay: "contacts" })
        this.context.updateActiveChat(null)
    }

    sendMessage = (chatId, messageContent) => {
        const message = { chat_id: chatId, sender_id: this.context.user.id, message_content: messageContent }
        const { activeChat, user } = this.context;
        const receiver = activeChat.user1.id === user.id ? activeChat.user2 : activeChat.user1;
        ChatService.postMessage(message)
            .then(msg => {
                this.context.addNewMessage(msg, chatId)
                this.context.socket.emit(MESSAGE_SENT, { sender: user, receiverId: receiver.id, message: msg })
            })
    }

    sendTyping = (chatId, isTyping) => {
        const { socket } = this.context;
        socket.emit(TYPING, { chatId, isTyping })
    }
    
    render() {
        const { user, activeChat } = this.context;
        const { chats } = this.state;

        return (
            <div className={styles.container}>
                {(this.state.chats) && <>
                <MessageSideBar mobileDisplay={this.state.mobileDisplay === "contacts" ? true : false} chats={chats} user={user} activeChat={activeChat} setActiveChat={this.setActiveChat} />
                <div className={`${styles.chatRoomContainer} ${this.state.mobileDisplay === "chats" ? styles.activeMobile : styles.inactiveMobile}`}>
                    {
                        activeChat !== null ? (
                            <div className={styles.chatRoom}>
                                <MessageHeading mobileDisplayContacts={this.mobileDisplayContacts} chatId={activeChat.id} receiver={activeChat.user1.id === user.id ? activeChat.user2 : activeChat.user1}/>
                                <Messages messages={activeChat.messages} receiver={activeChat.user1.id === user.id ? activeChat.user2 : activeChat.user1} user={user} typingUsers={activeChat.typingUsers} />
                                <MessageInput sendMessage={message => this.sendMessage(activeChat.id, message)} sendTyping={isTyping => this.sendTyping(activeChat.id, isTyping)} />
                            </div>
                        )
                        : 
                        <div className={`${styles.chatRoom} ${styles.choose}`}>
                            <h3>Choose a chat!</h3>
                        </div>
                    }
                </div>
                </>}
            </div>
        )
    }
}