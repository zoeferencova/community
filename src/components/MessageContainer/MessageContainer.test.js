import React from "react";
import { render } from "enzyme";
import MessageContainer from "./MessageContainer";
import CommUnityContext from "../../contexts/context";
import mockContext from "../../contexts/mock-context";

describe("MessageContainer", () => {
    it("renders MessageContainer component given mock context data", () => {
        const wrapper = render(
            <CommUnityContext.Provider value={mockContext}>
                <MessageContainer />
            </CommUnityContext.Provider>
        );
        
        expect(wrapper).toMatchSnapshot();
    });
});