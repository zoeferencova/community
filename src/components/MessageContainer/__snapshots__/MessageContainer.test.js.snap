// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageContainer renders MessageContainer component given mock context data 1`] = `
<div
  class="container"
>
  <div
    class="sideBar inactiveMobile"
    id="side-bar"
  >
    <div
      class="heading"
    >
      <div
        class="chatName"
      >
        Chats
      </div>
    </div>
    <div
      class="users"
    />
  </div>
  <div
    class="chatRoomContainer inactiveMobile"
  >
    <div
      class="chatRoom"
    >
      <div
        class="headerContainer"
      >
        <div
          class="chatHeader"
        >
          <div
            class="userInfo"
          >
            <i
              class="fas fa-chevron-left chatBackArrow"
            />
            <div
              class="circle indigo headerPic"
              first_name="James"
            >
              J
            </div>
            <div
              class="userName"
            >
              James
            </div>
            <div
              class="status"
            >
              <div
                class="indicator"
              />
            </div>
          </div>
          <div
            class="options"
          >
            <i
              class="fas fa-info-circle"
            />
            <i
              class="fas fa-times"
            />
          </div>
        </div>
      </div>
      <div
        class="threadContainer"
      >
        <div
          class="thread"
        >
          <div
            class="messageContainer right lastInThread"
          >
            <div
              class="circle blue messagePic"
              first_name="Zoe"
            >
              Z
            </div>
            <div
              class="data"
            >
              <div
                class="message"
              >
                hi
              </div>
              <div
                class="time"
              >
                2:22 PM
              </div>
            </div>
          </div>
          <div
            class="messageContainer false lastInThread"
          >
            <div
              class="circle indigo messagePic"
              first_name="James"
            >
              J
            </div>
            <div
              class="data"
            >
              <div
                class="message"
              >
                hey
              </div>
              <div
                class="time"
              >
                2:23 PM
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="messageInput"
      >
        <form
          class="messageForm"
        >
          <input
            autocomplete="off"
            class="formControl"
            id="message"
            placeholder="Type a message..."
            type="text"
            value=""
          />
          <button
            class="send"
            disabled=""
            type="submit"
          >
            <img
              class="sendIcon"
              src="https://img.icons8.com/ios-glyphs/96/89a1fc/paper-plane.png"
            />
          </button>
        </form>
      </div>
    </div>
  </div>
</div>
`;
